docs/proposals/buildstrategy-steps-resources.md

```diff
@@ -0,0 +1,152 @@
+the `Build` instance. For example, if users define `500m` as CPU request, but the strategy consists of five steps, then the used
+resources will be `500m` multiplied by `5`. This of course, can have implications for billing.
+
+## Motivation
+
+For strategies with multiple steps like Buildpacks, not all the steps required the same amount of resources. For this multi-step strategies we
+want to identify the **Step** that will consume the highest resources and allow users to be able to modify that step(_container_) resources through the `Build`. In other words, the more resources you assign to that step, the **faster** the build will be.
+
+### Goals
+
+- Keep the `Build` API untouched. Basically the `spec.resources` remains the same, but this configuration will only apply to one
+particular step in the related Build/ClusterBuildStrategy strategies.
+
+- Keep the user away from the strategies insight. User should only know that by adding more resources for X strategy, the build will be faster.
+
+- Allow users to modify one single Build/ClusterBuildStrategy strategies step(_the one that consumes the more resources_), by using the `Build` `spec.resources`.
+
+- Allow Build/ClusterBuildStrategy **admins** to define which step on the strategy will get the resources, by setting up a boolean flag. But also admins currently could
+  define particular resources for each step.
+
+### Non-Goals
+
+- Defining the concrete numbers for the resources(_e.g. CPU/memory_) used per steps in the different Build/ClusterBuildStrategy strategies.
+
+## Proposal
+
+Currently we assign the numbers inside the Build `spec.resources` into all the `steps` for the related strategy. We want to modify this
+by only applying those resources to a single step(_container_). The mechanism for doing this will consist of three parts:
+
+1. We will introduce a new field under the `BuildStep` struct, see `buildstrategy_types.go`. This field will be of the type `boolean`. We propose the following:
+   ```go
+    // BuildStep defines a partial step that needs to run in container for
+    // building the image.
+    type BuildStep struct {
+    	ResourceEnabled  bool `yaml:"resourceEnable,omitempty"`
+    	corev1.Container `yaml:",inline"`
+    }
+   ```
+2. The Build/ClusterBuildStrategy admin will be able to set the above field via a key on the step he/she considers is the more resource consumer.
+   ```yaml
+    - name: step-build
+      image: $(build.builder.image)
+      resourceEnable: true
+      securityContext:
+        runAsUser: 1000
+      command:
+        - /cnb/lifecycle/builder
+      args:
+        - -app=/workspace/source/$(build.source.contextDir)
+        - -layers=/layers
+        - -group=/layers/group.toml
+        - -plan=/layers/plan.toml
+      volumeMounts:
+        - name: layers-dir
+          mountPath: /layers
+   ```
+3. We need to use the above flag in the logic, inside the `generate_taskrun_test.go`, during the loop into all of the strategy steps.
```

Comment:
What are we going to do if multiple steps in a build run have the flag set to `true`?

- reject this (not desired as I understand it based on later comments)
- set the resource on multiple steps (means that what the user defines is not what he gets)
- devide the resource by the number of steps with the flag set to `true` and apply those values

Reply:
To answer your question, `set the resource on multiple steps`. This proposal is mainly to fix the current flaw we have where we actually apply `spec.resources` to all containers(steps) in a strategy. Where basically the user can get more resources than what he/she requested.

In this above flaw, we already have the problem of `what the user defines is not what he gets`.

For the problem of `what the user defines is not what he gets`, this proposal is not fixing that. To fix that, we definitely need to modify `spec.resources`, while it currently follow an 1 -> N(strategies steps) relationship, and I think we are looking for a N -> N.

Reply:
Also, even when the flag is set to `true` for a single container, Tekton might use the namespace `LimitRange` minimum values(cpu/mem) and assign those to the rest of the steps that didn´t defined resources, meaning `what the user defines is not what he gets` .This is maybe a broader problem, that could be only tackle by having a new API field in the BuildRun with a conglomerate from all the steps resources, calculated during runtime.